// typed-rpc/dist/client.mjs
function isJsonRpcResponse(res) {
  if (typeof res !== "object" || res === null) return false;
  if (!("jsonrpc" in res) || res.jsonrpc !== "2.0") return false;
  if (!("id" in res) || typeof res.id !== "string" && typeof res.id !== "number" && res.id !== null)
    return false;
  if ("result" in res) {
    return !("error" in res);
  } else if ("error" in res) {
    const error = res.error;
    return typeof error === "object" && error !== null && "code" in error && typeof error.code === "number" && "message" in error && typeof error.message === "string";
  }
  return false;
}
class RpcError extends Error {
  code;
  data;
  constructor(message, code, data) {
    super(message);
    this.name = "RpcError";
    this.code = code;
    this.data = data;
    Object.setPrototypeOf(this, RpcError.prototype);
  }
}
const identityTranscoder = {
  serialize: (data) => data,
  deserialize: (data) => data
};
function rpcClient(options) {
  let transport;
  let transcoder = identityTranscoder;
  let uuid;
  if (typeof options === "string") {
    transport = fetchTransport({ url: options });
  } else if ("transport" in options && options.transport) {
    transport = options.transport;
    transcoder = options.transcoder || identityTranscoder;
    uuid = options.uuid;
  } else {
    transport = fetchTransport(options);
    transcoder = options.transcoder || identityTranscoder;
    uuid = options.uuid;
  }
  const { serialize, deserialize } = transcoder;
  const sendRequest = async (method, args, signal) => {
    const req = createRequest(method, args, uuid);
    const raw = await transport(serialize(req), signal);
    const res = deserialize(raw);
    if (!isJsonRpcResponse(res)) {
      throw new TypeError("Not a JSON-RPC 2.0 response");
    }
    if ("error" in res) {
      const { code, message, data } = res.error;
      throw new RpcError(message, code, data);
    } else {
      return res.result;
    }
  };
  const abortControllers = /* @__PURE__ */ new WeakMap();
  const target = {
    /**
     * Abort the request for the given promise.
     */
    $abort: (promise) => {
      const ac = abortControllers.get(promise);
      ac?.abort();
    }
  };
  return new Proxy(target, {
    /* istanbul ignore next */
    get(target2, prop, receiver) {
      if (Reflect.has(target2, prop)) {
        return Reflect.get(target2, prop, receiver);
      }
      if (typeof prop === "symbol") return;
      if (prop === "then") return;
      if (prop === "toJSON") return;
      return (...args) => {
        const ac = new AbortController();
        const promise = sendRequest(prop.toString(), args, ac.signal);
        abortControllers.set(promise, ac);
        promise.finally(() => {
          abortControllers.delete(promise);
        }).catch(() => {
        });
        return promise;
      };
    }
  });
}
function createRequest(method, params, uuid) {
  const req = {
    jsonrpc: "2.0",
    id: uuid ? uuid() : Date.now().toString(36) + Math.random().toString(36).substring(2),
    method
  };
  if (params?.length) {
    req.params = removeTrailingUndefs(params);
  }
  return req;
}
function removeTrailingUndefs(values) {
  const a = [...values];
  while (a.length && a[a.length - 1] === void 0) a.length--;
  return a;
}
function fetchTransport(options) {
  return async (req, signal) => {
    const headers = options?.getHeaders ? await options.getHeaders() : {};
    const res = await fetch(options.url, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        ...headers
      },
      body: JSON.stringify(req),
      credentials: options?.credentials,
      signal
    });
    if (!res.ok) {
      throw new RpcError(res.statusText, res.status);
    }
    return await res.json();
  };
}

export { RpcError, createRequest, fetchTransport, isJsonRpcResponse, removeTrailingUndefs, rpcClient };
