// typed-rpc/dist/server.mjs
function isJsonRpcRequest(req) {
  if (typeof req !== "object" || req === null) return false;
  const request = req;
  return request.jsonrpc === "2.0" && typeof request.method === "string" && (request.id === void 0 || typeof request.id === "string" || typeof request.id === "number" || request.id === null) && (request.params === void 0 || Array.isArray(request.params));
}
function hasProperty(obj, prop) {
  return typeof obj === "object" && obj !== null && prop in obj;
}
function hasMethod(obj, prop) {
  return hasProperty(obj, prop) && typeof obj[prop] === "function";
}
function getErrorCode(err) {
  if (hasProperty(err, "code") && typeof err.code === "number") {
    return err.code;
  }
  return -32e3;
}
function getErrorMessage(err) {
  if (hasProperty(err, "message") && typeof err.message === "string") {
    return err.message;
  }
  return "";
}
function getErrorData(err) {
  if (hasProperty(err, "data")) {
    const stringifiedData = JSON.stringify(err.data);
    if (stringifiedData !== void 0) {
      err.data = JSON.parse(stringifiedData);
    }
    return err.data;
  }
}
function getRequestId(req) {
  if (hasProperty(req, "id")) {
    const id = req.id;
    if (typeof id === "string" || typeof id === "number") return id;
  }
  return null;
}
async function handleRpc(request, service, options) {
  const req = options?.transcoder?.deserialize(request) ?? request;
  const id = getRequestId(req);
  const res = (data) => {
    const raw = {
      jsonrpc: "2.0",
      id,
      ...data
    };
    return options?.transcoder?.serialize(raw) ?? raw;
  };
  if (!isJsonRpcRequest(req)) {
    return res({ error: { code: -32600, message: "Invalid Request" } });
  }
  const { method, params } = req;
  if (!hasMethod(service, method)) {
    return res({
      error: { code: -32601, message: `Method not found: ${method}` }
    });
  }
  try {
    const result = await service[method](...params ?? []);
    return res({ result });
  } catch (err) {
    if (options?.onError) {
      options.onError(err);
    }
    return res({
      error: {
        code: (options?.getErrorCode ?? getErrorCode)(err),
        message: (options?.getErrorMessage ?? getErrorMessage)(err),
        data: (options?.getErrorData ?? getErrorData)(err)
      }
    });
  }
}

export { handleRpc, isJsonRpcRequest };
