// plainjob/dist/plainjob.js
// src/queue.ts
import cronParser from "cron-parser";

// src/jobs.ts
var JobStatus = /* @__PURE__ */ ((JobStatus2) => {
  JobStatus2[JobStatus2["Pending"] = 0] = "Pending";
  JobStatus2[JobStatus2["Processing"] = 1] = "Processing";
  JobStatus2[JobStatus2["Done"] = 2] = "Done";
  JobStatus2[JobStatus2["Failed"] = 3] = "Failed";
  return JobStatus2;
})(JobStatus || {});
var ScheduledJobStatus = /* @__PURE__ */ ((ScheduledJobStatus2) => {
  ScheduledJobStatus2[ScheduledJobStatus2["Idle"] = 0] = "Idle";
  ScheduledJobStatus2[ScheduledJobStatus2["Processing"] = 1] = "Processing";
  return ScheduledJobStatus2;
})(ScheduledJobStatus || {});

// src/queue.ts
function better(database) {
  return {
    driver: "better-sqlite3",
    filename: database.name,
    pragma: database.pragma.bind(database),
    transaction: database.transaction.bind(database),
    exec: database.exec.bind(database),
    prepare: database.prepare.bind(database),
    close: database.close.bind(database)
  };
}
function bun(database) {
  return {
    driver: "bun:sqlite",
    filename: database.filename,
    pragma: (stmt) => {
      database.exec(`PRAGMA ${stmt};`);
    },
    transaction: database.transaction.bind(database),
    exec: database.exec.bind(database),
    prepare: database.query.bind(database),
    close: database.close.bind(database)
  };
}
function defineQueue(opts) {
  const db = opts.connection;
  const log = opts.logger || console;
  const jobProcessingTimeout = opts.timeout || 30 * 60 * 1e3;
  const maintenanceInterval = opts.maintenanceInterval || 60 * 1e3;
  const removeDoneJobsOlderThan = opts.removeDoneJobsOlderThan || 7 * 24 * 60 * 60 * 1e3;
  const removeFailedJobsOlderThan = opts.removeFailedJobsOlderThan || 30 * 24 * 60 * 60 * 1e3;
  let maintenanceTimeout;
  const serializer = opts.serializer || ((data) => JSON.stringify(data));
  db.pragma("journal_mode = WAL");
  db.pragma("synchronous = 1");
  db.pragma("busy_timeout = 5000");
  db.exec(`
    CREATE TABLE IF NOT EXISTS plainjob_jobs (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      type TEXT NOT NULL,
      data TEXT NOT NULL,
      status INTEGER DEFAULT 0 NOT NULL,
      failed_at INTEGER,
      error TEXT,
      next_run_at INTEGER,
      created_at INTEGER NOT NULL
    );

    CREATE INDEX IF NOT EXISTS idx_jobs_status_type_next_run_at ON plainjob_jobs (status, type, next_run_at);

    CREATE TABLE IF NOT EXISTS plainjob_scheduled_jobs (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      type TEXT NOT NULL UNIQUE,
      status INTEGER DEFAULT 0 NOT NULL,
      cron_expression TEXT,
      next_run_at INTEGER,
      created_at INTEGER NOT NULL
    );

    CREATE INDEX IF NOT EXISTS idx_scheduled_jobs_status_type_next_run_at ON plainjob_scheduled_jobs (status, type, next_run_at);
  `);
  const removeDoneJobsStmt = db.prepare(`
    DELETE FROM plainjob_jobs 
    WHERE status = ${2 /* Done */} AND next_run_at < @threshold
  `);
  const removeFailedJobsStmt = db.prepare(`
    DELETE FROM plainjob_jobs 
    WHERE status = ${3 /* Failed */} AND failed_at < @threshold
  `);
  function initializeMaintenance() {
    maintenanceTimeout = setInterval(() => {
      queue.requeueTimedOutJobs(jobProcessingTimeout);
      queue.removeDoneJobs(removeDoneJobsOlderThan);
      queue.removeFailedJobs(removeFailedJobsOlderThan);
    }, maintenanceInterval);
  }
  const countJobsStmt = db.prepare("SELECT COUNT(*) FROM plainjob_jobs");
  const countJobsByTypeAndStatusStmt = db.prepare(
    "SELECT COUNT(*) FROM plainjob_jobs WHERE status = @status AND type = @type"
  );
  const countJobsByStatusStmt = db.prepare(
    "SELECT COUNT(*) FROM plainjob_jobs WHERE status = @status"
  );
  const countJobsByTypeStmt = db.prepare(
    "SELECT COUNT(*) FROM plainjob_jobs WHERE type = @type"
  );
  const getJobTypesStmt = db.prepare("SELECT DISTINCT type FROM plainjob_jobs");
  const getJobByIdStmt = db.prepare(`
    SELECT 
      id,
      type,
      data,
      status,
      created_at as createdAt,
      failed_at as failedAt,
      next_run_at as nextRunAt,
      error
    FROM plainjob_jobs 
    WHERE id = ?
  `);
  const getScheduledJobByIdStmt = db.prepare(`
    SELECT 
      id,
      type,
      status,
      created_at as createdAt,
      cron_expression as cronExpression,
      next_run_at as nextRunAt
    FROM plainjob_scheduled_jobs 
    WHERE id = ?
  `);
  const getScheduledJobByTypeStmt = db.prepare(`
    SELECT 
      id,
      type,
      status,
      created_at as createdAt,
      cron_expression as cronExpression,
      next_run_at as nextRunAt
    FROM plainjob_scheduled_jobs 
    WHERE type = @type
  `);
  const getScheduledJobsStmt = db.prepare(`
    SELECT 
      id,
      type,
      status,
      created_at as createdAt,
      cron_expression as cronExpression,
      next_run_at as nextRunAt
    FROM plainjob_scheduled_jobs
    ORDER BY created_at
  `);
  const insertJobStmt = db.prepare(
    "INSERT INTO plainjob_jobs (type, data, created_at, next_run_at) VALUES (@type, @data, @createdAt, @nextRunAt)"
  );
  const insertScheduledJobStmt = db.prepare(
    "INSERT INTO plainjob_scheduled_jobs (type, cron_expression, next_run_at, created_at) VALUES (@type, @cronExpression, @nextRunAt, @createdAt)"
  );
  const updateScheduledJobCronExpressionStmt = db.prepare(`
    UPDATE plainjob_scheduled_jobs SET cron_expression = @cronExpression WHERE id = @id
  `);
  const requeueTimedOutJobsStmt = db.prepare(`
    UPDATE plainjob_jobs SET status = ${0 /* Pending */} WHERE status = ${1 /* Processing */} AND next_run_at < @threshold
  `);
  const getAndMarkJobAsProcessingStmt = db.prepare(`
  UPDATE plainjob_jobs SET status = ${1 /* Processing */}
  WHERE id = (
    SELECT id FROM plainjob_jobs
    WHERE status = ${0 /* Pending */} AND type = @type AND next_run_at <= @now
    ORDER BY next_run_at LIMIT 1
  )
`);
  const getUpdatedJobStmt = db.prepare(`
  SELECT * FROM plainjob_jobs
  WHERE status = ${1 /* Processing */} AND type = @type AND next_run_at <= @now
  ORDER BY next_run_at LIMIT 1
`);
  const getJobIdToProcessNextStmt = db.prepare(`
  SELECT id
  FROM plainjob_jobs 
  WHERE status = ${0 /* Pending */} AND type = @type AND next_run_at <= @now
  ORDER BY next_run_at LIMIT 1
`);
  const updateJobAsProcessingStmt = db.prepare(`
  UPDATE plainjob_jobs SET status = ${1 /* Processing */}
  WHERE id = @id
`);
  const getNextScheduledJobStmt = db.prepare(`
    SELECT 
      id,
      type,
      status,
      created_at as createdAt,
      cron_expression as cronExpression,
      next_run_at as nextRunAt
    FROM plainjob_scheduled_jobs 
    WHERE status = ${0 /* Idle */} AND next_run_at <= @now
    ORDER BY next_run_at LIMIT 1
  `);
  const updateJobStatusStmt = db.prepare(`
    UPDATE plainjob_jobs SET status = @status WHERE id = @id
  `);
  const updateScheduledJobStatusStmt = db.prepare(`
    UPDATE plainjob_scheduled_jobs SET status = @status, next_run_at = @nextRunAt WHERE id = @id
  `);
  const failJobStmt = db.prepare(`
    UPDATE plainjob_jobs SET status = ${3 /* Failed */}, failed_at = @failedAt, error = @error WHERE id = @id
  `);
  const queue = {
    add(type, data, options = {}) {
      const now = Date.now();
      const serializedData = serializer(data);
      const result = insertJobStmt.run({
        type,
        data: serializedData,
        createdAt: now,
        nextRunAt: now + (options.delay ?? 0)
      });
      return { id: result.lastInsertRowid };
    },
    addMany(type, dataList, options = {}) {
      const now = Date.now();
      const ids = [];
      const insertManyJobs = db.transaction(
        (jobs2) => {
          for (const job of jobs2) {
            const result = insertJobStmt.run(job);
            ids.push(result.lastInsertRowid);
          }
        }
      );
      const jobs = dataList.map((data) => ({
        type,
        data: serializer(data),
        createdAt: now,
        nextRunAt: now + (options.delay ?? 0)
      }));
      insertManyJobs(jobs);
      return { ids };
    },
    schedule(type, { cron }) {
      try {
        cronParser.parseExpression(cron);
      } catch (error) {
        throw new Error(
          `invalid cron expression provided: ${cron} ${error.message}`
        );
      }
      const found = getScheduledJobByTypeStmt.get({
        type
      }) ?? void 0;
      if (found) {
        log.debug(
          `updating existing scheduled job ${found.id} with cron expression ${cron}`
        );
        updateScheduledJobCronExpressionStmt.run({
          cronExpression: cron,
          id: found.id
        });
        return { id: found.id };
      }
      const result = insertScheduledJobStmt.run({
        type,
        cronExpression: cron,
        nextRunAt: 0,
        createdAt: Date.now()
      });
      return { id: result.lastInsertRowid };
    },
    countJobs(opts2) {
      if (opts2?.type && opts2?.status !== void 0) {
        const result2 = countJobsByTypeAndStatusStmt.get({
          type: opts2.type,
          status: opts2.status
        }) ?? void 0;
        return result2["COUNT(*)"];
      }
      if (opts2?.status !== void 0 && !opts2?.type) {
        const result2 = countJobsByStatusStmt.get({ status: opts2.status }) ?? void 0;
        return result2["COUNT(*)"];
      }
      if (opts2?.type && opts2?.status === void 0) {
        const result2 = countJobsByTypeStmt.get({ type: opts2.type }) ?? void 0;
        return result2["COUNT(*)"];
      }
      const result = countJobsStmt.get() ?? void 0;
      return result["COUNT(*)"];
    },
    getJobById(id) {
      return getJobByIdStmt.get(id) ?? void 0;
    },
    getJobTypes() {
      const result = getJobTypesStmt.all();
      return result.map((row) => row.type);
    },
    getScheduledJobs() {
      const result = getScheduledJobsStmt.all();
      return result;
    },
    getScheduledJobById(id) {
      return getScheduledJobByIdStmt.get(id);
    },
    requeueTimedOutJobs(timeout) {
      const now = Date.now();
      const result = requeueTimedOutJobsStmt.run({ threshold: now - timeout });
      log.debug(`requeued ${result.changes} timed out jobs`);
      if (opts.onProcessingJobsRequeued) {
        opts.onProcessingJobsRequeued(result.changes);
      }
    },
    removeDoneJobs(olderThan) {
      const now = Date.now();
      const result = removeDoneJobsStmt.run({ threshold: now - olderThan });
      log.debug(`removed ${result.changes} done jobs`);
      if (opts.onDoneJobsRemoved) {
        opts.onDoneJobsRemoved(result.changes);
      }
    },
    removeFailedJobs(olderThan) {
      const now = Date.now();
      const result = removeFailedJobsStmt.run({ threshold: now - olderThan });
      log.debug(`removed ${result.changes} failed jobs`);
      if (opts.onFailedJobsRemoved) {
        opts.onFailedJobsRemoved(result.changes);
      }
    },
    getAndMarkJobAsProcessing(type) {
      const now = Date.now();
      return db.transaction(() => {
        const id = getJobIdToProcessNextStmt.get({ type, now });
        if (!id || !id.id) return void 0;
        updateJobAsProcessingStmt.run({ id: id.id });
        return id;
      }).immediate();
    },
    getAndMarkScheduledJobAsProcessing() {
      return db.transaction(() => {
        const job = getNextScheduledJobStmt.get({
          now: Date.now()
        }) ?? void 0;
        if (job) {
          updateScheduledJobStatusStmt.run({
            status: 1 /* Processing */,
            id: job.id,
            nextRunAt: job.nextRunAt
          });
          return { ...job, status: 1 /* Processing */ };
        }
        return void 0;
      }).immediate();
    },
    markJobAsDone(id) {
      return updateJobStatusStmt.run({ status: 2 /* Done */, id });
    },
    markJobAsFailed(id, error) {
      return failJobStmt.run({
        failedAt: Date.now(),
        error,
        id
      });
    },
    markScheduledJobAsIdle(id, nextRunAt) {
      return updateScheduledJobStatusStmt.run({
        status: 0 /* Idle */,
        id,
        nextRunAt
      });
    },
    close() {
      if (maintenanceTimeout) {
        clearInterval(maintenanceTimeout);
      }
      db.close();
    }
  };
  initializeMaintenance();
  return queue;
}

// src/worker.ts
import cronParser2 from "cron-parser";
function defineWorker(jobType, processor, options) {
  const log = options.logger || console;
  const id = Math.random().toString(36).substring(2, 15);
  const pollInterval = options.pollIntervall ?? 1e3;
  const queue = options.queue;
  let shouldKeepRunning = false;
  let isRunning = true;
  let cancelSleep;
  let sleeping;
  async function processScheduledJobs() {
    log.debug(`worker [${id}] checking for scheduled jobs`);
    const scheduledJob = queue.getAndMarkScheduledJobAsProcessing();
    if (scheduledJob) {
      log.debug(
        `worker [${id}] processing scheduled job '${scheduledJob.id}' '${scheduledJob.type}'`
      );
      const nextRunAt = cronParser2.parseExpression(scheduledJob.cronExpression).next().toDate().getTime();
      queue.markScheduledJobAsIdle(scheduledJob.id, nextRunAt);
      log.debug(
        `worker [${id}] marking scheduled job ${scheduledJob.id} as 'idle'`
      );
      queue.add(scheduledJob.type, {});
      log.debug(
        `worker [${id}] adding job '${scheduledJob.id}' '${scheduledJob.type}' from scheduled job`
      );
      return true;
    }
    return false;
  }
  async function processRegularJobs() {
    log.debug(`worker [${id}] checking for regular jobs`);
    const jobId = queue.getAndMarkJobAsProcessing(jobType);
    if (jobId) {
      const job = queue.getJobById(jobId.id);
      if (options.onProcessing) {
        const job2 = queue.getJobById(jobId.id);
        options.onProcessing(job2);
      }
      log.debug(
        `worker [${id}] processing job ${job.id}, ${job.type}, ${job.data}`
      );
      try {
        await processor({ id: job.id, data: job.data, type: job.type });
        queue.markJobAsDone(job.id);
        if (options.onCompleted) {
          options.onCompleted(job);
        }
        log.debug(`worker [${id}] marking job ${job.id} as 'done'`);
      } catch (error) {
        const errorMessage = `${error.stack}
${error.message}`;
        queue.markJobAsFailed(job.id, errorMessage);
        if (options.onFailed) {
          options.onFailed(job, errorMessage);
        }
        log.error(`worker [${id}] marking job ${job.id} as 'failed'`);
      }
      return true;
    }
    return false;
  }
  async function start() {
    try {
      shouldKeepRunning = true;
      while (shouldKeepRunning) {
        const processedScheduled = await processScheduledJobs();
        const processedRegular = await processRegularJobs();
        if (!processedScheduled && !processedRegular && isRunning && shouldKeepRunning) {
          sleeping = new Promise((resolve) => {
            const timeout = setTimeout(resolve, pollInterval);
            cancelSleep = () => {
              clearTimeout(timeout);
              resolve();
            };
          });
          await sleeping;
          cancelSleep = void 0;
        }
      }
      isRunning = false;
    } catch (error) {
      log.error(`worker [${id}] encountered error: ${error}`);
      throw error;
    }
  }
  async function stop() {
    log.info(`worker [${id}] shutting down...`);
    shouldKeepRunning = false;
    if (cancelSleep) {
      cancelSleep();
    }
    log.debug(`worker [${id}] waiting for worker to stop...`);
    await sleeping;
    log.debug(`worker [${id}] shut down`);
  }
  return {
    start,
    stop
  };
}
export {
  JobStatus,
  ScheduledJobStatus,
  better,
  bun,
  defineQueue,
  defineWorker
};
